File: Chapter 03/C#/Section_3_4_OrderProcessor_Refactor/After/OrderService.cs
namespace Chapter03.OrderProcessorRefactor.After;

/// <summary>
/// THE SOLUTION: The Coordinator / Facade.
/// 
/// ARCHITECTURE NOTE: Look closely at this class. Notice what it DOESN'T do:
/// It doesn't validate, it doesn't charge credit cards, and it doesn't send emails. 
/// 
/// Instead, it acts as a high-level "Coordinator". It orchestrates the flow of 
/// the transaction by delegating the actual work to the injected services. 
/// 
/// By using Constructor Injection (DIP), we can easily pass in "Mock" versions 
/// of the PaymentService and NotificationService to completely unit test this 
/// entire checkout flow in milliseconds without ever hitting a real database 
/// or charging a real credit card!
/// </summary>
public class OrderService
{
    // Dependencies are explicitly declared.
    private readonly OrderValidator _validator;
    private readonly PaymentService _paymentService;
    private readonly InventoryManager _inventoryManager;
    private readonly NotificationService _notificationService;

    /// <summary>
    /// Dependencies are injected from the outside (Constructor Injection).
    /// </summary>
    public OrderService(
       OrderValidator validator, PaymentService payment, 
       InventoryManager inventory, NotificationService notifier)
    {
        _validator = validator;
        _paymentService = payment;
        _inventoryManager = inventory;
        _notificationService = notifier;
    }

    /// <summary>
    /// The high-level transaction script is now clean, readable, and safe.
    /// </summary>
    public string ProcessOrder(Order order)
    {
        _validator.Validate(order);

        if (_paymentService.ProcessPayment(order))
        {
            _inventoryManager.UpdateInventory(order);
            _notificationService.SendConfirmationEmail(order);
            return "Order processed successfully.";
        }
        else
        {
            return "Payment failed.";
        }
    }
}

File: Chapter 03/JS/section_3_4_orderProcessor_refactor/after/orderService.js
/**
 * THE SOLUTION: The Coordinator / Facade.
 * * ARCHITECTURE NOTE: This class no longer contains "how-to" logic; it only 
 * contains "when-to" logic. 
 * * By using Constructor Injection, we've achieved full Dependency Inversion (DIP). 
 * This class doesn't care HOW an order is validated or HOW a payment is 
 * processed; it simply coordinates the flow between its injected dependencies.
 */
class OrderService {
    /**
     * Dependencies are injected from the outside.
     */
    constructor(validator, paymentService, inventoryManager, notificationService) {
        this.validator = validator;
        this.paymentService = paymentService;
        this.inventoryManager = inventoryManager;
        this.notificationService = notificationService;
    }

    /**
     * Orchestrates the high-level business process.
     */
    processOrder(order) {
        // Step 1: Delegate validation
        this.validator.validate(order);

        // Step 2: Orchestrate the successful path
        if (this.paymentService.processPayment(order)) {
            this.inventoryManager.updateInventory(order);
            this.notificationService.sendConfirmationEmail(order);
            return "Order processed successfully.";
        } else {
            return "Payment failed.";
        }
    }
}

module.exports = OrderService;